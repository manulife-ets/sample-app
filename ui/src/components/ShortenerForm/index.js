import React, { useState } from 'react';
import Grid from '@radd/react/dist/components/containers/Grid';
import { TextField } from '../TextField';
import { Button } from '../Button';

const fetcher = async ({url, method, body}) => {
  return fetch(url, {
    method: method,
    headers: {
      'Content-Type': 'application/json',
    },
    body: body
  });
}

const ShortenerForm = ({task}) => {
  const [longUrl, setLongUrl] = useState('');
  const [urlId, setUrlId] = useState('');
  const [shortUrl, setShortUrl] = useState('');
  const [errorMsg, setErrorMsg] = useState('');

  const updateUrl = e => {
    setLongUrl(e.target.value);
    setErrorMsg(false);
  }

  const updateUrlId = e => {
    setUrlId(e.target.value);
  }

  const handleSubmit = e => {
    e.preventDefault();
    processShortenRequest(longUrl);
  }

  const processShortenRequest = async str => {
    const url = isValidUrl(str) && str;
    if (str.length < 1) {
      renderError('EMPTY_INPUT');
      return false;
    }
    if (!url) {
      renderError('INVALID_INPUT');
      return false;
    } else {
      await makeShortenUrlRequest(url);
    }
  }

  const isValidUrl = str => {
    const urlPattern = new RegExp(/^https?:\/\/(?:\([-A-Z0-9+&@#\/%=~_|$?!:,.]*\)|[-A-Z0-9+&@#\/%=~_|$?!:,.])*(?:\([-A-Z0-9+&@#\/%=~_|$?!:,.]*\)|[A-Z0-9+&@#\/%=~_|$])$/, 'i');
    return !!urlPattern.test(str);
  }

  const buildRequestFromTask = (task, url) => {
    switch (task) {
      case 'SHORTEN':
        return {
          url: '/api/v2',
          method: 'POST',
          body: JSON.stringify({
            longURL: url
          })
        }
      case 'UPDATE':
        return {
          url: '/api/v2/' + urlId,
          method: 'PUT',
          body: JSON.stringify({
            longURL: url
          })
        }
      case 'CREATE':
        return {
          url: '/api/v2',
          method: 'POST',
          body: JSON.stringify({
            id: urlId,
            longURL: url
          })
        }
      default: renderError('EMPTY')
    }

  }

  const makeShortenUrlRequest = async url => {
    let shortUrl;
    const requestParams = buildRequestFromTask(task, url);
    try {
      let response = await fetcher(requestParams);
      let json = await response.json()
      if (!response.ok) {
        renderError('API_ERROR', json.message)
        return false;
      } else {
        shortUrl = json.id;
        setShortUrl(shortUrl);
        document.getElementById('shortUrl').select();
      }
    } catch(err) {
      renderError('API_ERROR', 'URL ID not found');
    }
  }

  const handleCopy = e => {
    e.preventDefault();
    copyShortUrl();
  }

  const copyShortUrl = () => {
    const shortUrl = document.getElementById('shortUrl');
    shortUrl.select();
    document.execCommand('copy');
  }

  const renderError = (type, message = 'Error calling API') => {
    const errorTypes = new Map([
      ['EMPTY_INPUT', 'Please enter a URL'],
      ['INVALID_INPUT', 'URL is not valid'],
      ['API_ERROR', message],
      ['EMPTY', '&nbsp;']
    ])
    setErrorMsg(errorTypes.get(type));
  }

  const placeholder = task === 'CREATE' ? 'Like a-custom-link-name or ✔🙌👍' : 'ex: custom-link-name or oqw3k78oz51m83y';
  const slogans = ['do-the-right-thing', 'think-big', 'get-it-done-together', 'own-it', 'share-your-humanity', 'obsess-about-customers'];
  const slogan = slogans[Math.floor(Math.random() * slogans.length)];
  // use dev api if running locally
  const apiUrl = window.location.origin.indexOf('localhost') > 0 ? 'https://go-dev.manulife.com' : window.location.origin;
  return (
  <>
    <form onSubmit={handleSubmit} autoComplete="off">
        {((task === 'CREATE') || (task === 'UPDATE')) &&
        <Grid xs="3fr 1fr" gutter="0" margin="2rem none 2rem">
          <TextField
            value={urlId}
            inputLabel="Enter a URL ID"
            name="urlId"
            placeholder={placeholder}
            onChange={updateUrlId}
          />
        </Grid>
      }
      <Grid xs="3fr 1fr" gutter="0" margin="none none 2rem">
        <TextField
          value={longUrl}
          inputLabel="Enter a looooong URL"
          name="longUrl"
            onChange={updateUrl}
            placeholder={`ex: https://www.manulife.com/really-really-really-really-really-long-url-to-${slogan}`}
          error={errorMsg}
        />
        <Button>{task.charAt(0).toUpperCase() + task.slice(1).toLowerCase()} It!</Button>
      </Grid>
    </form>
    {shortUrl &&
      <form onSubmit={handleCopy}>
        <Grid xs="3fr 1fr" gutter="0">
          <TextField
            value={`${apiUrl}/${shortUrl}`}
            inputLabel="Now copy that URL"
            name="shortUrl"
            readOnly={true}
          />
          <Button bgColor="#06c7ba" hoverColor="#08a298">Copy It!</Button>
        </Grid>
      </form>
    }
  </>
  );
}

export default ShortenerForm;
