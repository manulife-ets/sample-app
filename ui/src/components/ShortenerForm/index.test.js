import React from 'react'
import { render, fireEvent, findByText } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import ShortenerForm from '.';
import { act } from 'react-dom/test-utils';

describe('ShortenerForm', () => {
  test('it works', () => {
    const { container } = render(<ShortenerForm task='TEST' />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('it validates the URL', async () => {
    const { container, getByLabelText, getByText, findByRole } = render(<ShortenerForm task='TEST' />);
    const handleSubmit = jest.fn()

    fireEvent.change(getByLabelText(/enter a lo+ng url/i), { target: { value: 'htw/:qw.c/' } });
    fireEvent.click(getByText(/test it/i));

    const alert = await findByRole('alert');
    expect(alert).toHaveTextContent(/url is not valid/i);
  });

  test('it requires a URL to be entered', async () => {
    const { container, getByLabelText, getByText, findByRole } = render(<ShortenerForm task='TEST' />);

    fireEvent.change(getByLabelText(/enter a lo+ng url/i), { target: { value: '' } });
    fireEvent.click(getByText(/test it/i));

    const alert = await findByRole('alert');
    expect(alert).toHaveTextContent(/please enter a url/i);
  });

  test('it clears the error message when the input changes', async () => {
    const { container, getByLabelText, getByText, findByRole } = render(<ShortenerForm task='TEST' />);

    fireEvent.change(getByLabelText(/enter a lo+ng url/i), { target: { value: 'errrorrr cawzer' } });
    fireEvent.click(getByText(/test it/i));

    let alert = await findByRole('alert');
    expect(alert).toBeInTheDocument();

    fireEvent.change(getByLabelText(/enter a lo+ng url/i), { target: { value: 'errrorrr caw' } });

    alert = await document.querySelector('[role="alert"]');
    expect(alert).not.toBeInTheDocument();

  });

  test('it allows a valid URL', async () => {
    // mock out window.fetch for the test
    const makeShortenUrlRequest = jest.fn();
    const { container, getByLabelText, getByText } = render(<ShortenerForm task='SHORTEN' />);

    fireEvent.change(getByLabelText(/enter a lo+ng url/i), { target: { value: 'http://test.com' } });
    fireEvent.click(getByText(/shorten it/i));

    alert = await document.querySelector('[role="alert"]');
    expect(alert).not.toBeInTheDocument();
  });

  test('it renders the correct CREATE form', async () => {
    // mock out window.fetch for the test
    const makeShortenUrlRequest = jest.fn();
    const { container, getByLabelText, getByText } = render(<ShortenerForm task='CREATE' />);

    expect(getByText(/create it/i)).toBeInTheDocument();
  });

  test('it renders the correct CREATE form', async () => {
    // mock out window.fetch for the test
    const makeShortenUrlRequest = jest.fn();
    const { container, getByLabelText, getByText } = render(<ShortenerForm task='CREATE' />);

    expect(getByText(/create it/i)).toBeInTheDocument();
  });

  test('it renders the correct UPDATE form', async () => {
    // mock out window.fetch for the test
    const makeShortenUrlRequest = jest.fn();
    const { container, getByLabelText, getByText } = render(<ShortenerForm task='UPDATE' />);

    expect(getByText(/update it/i)).toBeInTheDocument();
  });

  test('it renders the API error message', async () => {
    global.fetch = jest.fn().mockImplementationOnce(() => {
      return new Promise((resolve, reject) => {
        resolve({
          ok: false,
          status: 400,
          json: () => { return { id: 'üÉè' } }
        });
      });
    });

    const { container, getByLabelText, getByText, findByRole } = render(<ShortenerForm task='TEST' />);

    fireEvent.change(getByLabelText(/enter a lo+ng url/i), { target: { value: 'errrorrr cawzer' } });
    fireEvent.click(getByText(/test it/i));

    const alert = await findByRole('alert');
    expect(alert).toBeInTheDocument();

  });

  test('it renders the copy URL form from the SHORTEN API response', async () => {
    global.fetch = jest.fn().mockImplementationOnce(() => {
      return new Promise((resolve, reject) => {
        resolve({
          ok: true,
          status: 200,
          json: () => { return { id: 'JESTTEST' } }
        });
      });
    });

    const { container, getByLabelText, getByText, findByText } = render(<ShortenerForm task='SHORTEN' />);

    fireEvent.change(getByLabelText(/enter a lo+ng url/i), { target: { value: 'http://www.com' } });
    fireEvent.click(getByText(/shorten it/i));

    const copy = await findByText('Copy It!');
    expect(copy).toBeInTheDocument();

    const input = getByLabelText('Now copy that URL');
    expect(input.value).toEqual('https://go.manulife.com/JESTTEST');
  });

  test('it renders the copy URL form from the CREATE API response', async () => {
    global.fetch = jest.fn().mockImplementationOnce(() => {
      return new Promise((resolve, reject) => {
        resolve({
          ok: true,
          status: 200,
          json: () => { return { id: 'JESTTEST' } }
        });
      });
    });

    const { container, getByLabelText, getByText, findByText } = render(<ShortenerForm task='CREATE' />);

    fireEvent.change(getByLabelText(/enter a url id/i), { target: { value: 'üÉè' } });

    fireEvent.change(getByLabelText(/enter a lo+ng url/i), { target: { value: 'http://www.com' } });
    fireEvent.click(getByText(/create it/i));

    const copy = await findByText('Copy It!');
    expect(copy).toBeInTheDocument();

    const input = getByLabelText('Now copy that URL');
    expect(input.value).toEqual('https://go.manulife.com/JESTTEST');
  });

  test('it renders the copy URL form from the CREATE API response', async () => {
    global.fetch = jest.fn().mockImplementationOnce(() => {
      return new Promise((resolve, reject) => {
        resolve({
          ok: true,
          status: 200,
          json: () => { return { id: 'üÉè' } }
        });
      });
    });

    const { container, getByLabelText, getByText, findByText } = render(<ShortenerForm task='UPDATE' />);

    fireEvent.change(getByLabelText(/enter a url id/i), { target: { value: 'üÉè' } });

    fireEvent.change(getByLabelText(/enter a lo+ng url/i), { target: { value: 'http://www.com' } });
    fireEvent.click(getByText(/update it/i));

    const copy = await findByText('Copy It!');
    expect(copy).toBeInTheDocument();

    const input = getByLabelText('Now copy that URL');
    expect(input.value).toEqual('https://go.manulife.com/üÉè');
  });
});
