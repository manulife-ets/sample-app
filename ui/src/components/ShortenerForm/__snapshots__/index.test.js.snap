// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShortenerForm it works 1`] = `
<form
  autocomplete="off"
>
  <div
    class="sc-bwzfXH bTvmza"
  >
    <div
      class="sc-htpNat dZrxDk"
    >
      <label
        class="sc-bxivhb dvLITu"
        for="longUrl"
      >
        Enter a looooong URL
      </label>
      <input
        class="sc-ifAKCX kHwQRe"
        id="longUrl"
        name="longUrl"
        placeholder="ex: https://www.manulife.com/really-really-really-really-really-long-url-to-get-it-done-together"
        type="text"
        value=""
      />
      
    </div>
    <button
      class="sc-bZQynM bmxWlX"
    >
      Test
       It!
    </button>
  </div>
</form>
`;
